apps/cave/components/AMM/Swap/CustomRecipient.tsx

```diff
@@ -11,21 +11,33 @@ import {
   useMultiStyleConfig,
 } from '@concave/ui'
 import { isAddress } from 'ethers/lib/utils'
-import { useReducer, useRef, useState } from 'react'
-import { useEnsAddress } from 'wagmi'
+import { useReducer, useRef } from 'react'
+import { usePrevious } from 'react-use'
+import { useDebounce } from 'usehooks-ts'
+import { Address, useEnsAddress } from 'wagmi'
 
-export const CustomRecipient = ({ onChangeRecipient }) => {
+export const CustomRecipient = ({
+  onChangeRecipient,
+}: {
+  onChangeRecipient: (r?: Address) => void
+}) => {
   const [isOpen, toggle] = useReducer((s) => !s, false)
 
-  const [recipient, setRecipient] = useState('')
+  const [recipient, setRecipient] = useReducer((_, r) => {
+    // if user types an address, set it as recipient
+    if (isAddress(r)) onChangeRecipient(r)
+    return r
+  }, '')
```

Comment:
kept it in a `useReducer` to make it easier to understand what's going on, the set recipient logic is together in the begining instead of a state here and two sets in the return down below (one for the internal state and another if it's a valid address callback for the parent)